### DISCUSSION TITLE
Serializing stuff & other shenanigans

### DISCUSSION THREAD

**Discussion by villares** [py5coding/py5generator#454] [Ideas]
Sometimes I miss the ability to serialize py5 objects that depend on Processing core objects inside.

Last year I played with making Py5Image objects work with Python's pickle, thanks to the `np_pixels` interface:

```python

import pickle
import py5

from py5 import Py5Image
old_new = Py5Image.__new__

def new_new(self, *args):
    if args:
        return old_new(self, *args)
    else:
        return object.__new__(Py5Image)   
    
def getstate(self):
    self.load_np_pixels()
    return self.np_pixels

def setstate(self, d):
    img = py5.create_image_from_numpy(d)
    self.__dict__ = img.__dict__

def setup():
    global img, unpickled_img
    py5.size(800, 800)
    # monkeypatching Py5Image
    Py5Image.__getstate__ = getstate
    Py5Image.__setstate__ = setstate
    Py5Image.__new__ = new_new
    
    img_path = 'sample.png'  # sample image file needed
    img = py5.load_image(img_path)
    
    unpickled_img = None
    
def draw():
    if unpickled_img is not None:
        py5.image(unpickled_img, 0, 0)

def key_pressed():
    if py5.key == 'd':
        with open('a.pickle', 'wb') as f:
            pickle.dump(img, f)
    elif py5.key == 'l':
        global unpickled_img
        with open('a.pickle', 'rb') as f:    
            unpickled_img = pickle.load(f)
            
py5.run_sketch(block=False)

```

Now consider this toy code:

```python
import py5

def setup():
    global shp
    py5.size(500, 500)
    shp = py5.create_shape(py5.ELLIPSE, 100, 100, 50, 50)

def draw():
    py5.background(100, 100, 200)
    py5.shape(shp, 0 ,0)

def key_pressed():
    global shp
    if py5.key == 's':
        save_shape(shp)
    if py5.key == 'l':
        shp = py5.load_shape('out.svg')
        
def save_shape(shp):
    with py5.begin_record(py5.SVG, 'out.svg') as r:
        r.shape(shp, 0, 0)
 
py5.run_sketch(block=False)
```
 
If `begin_record()` could accept a "file-like" buffer object instead of a string, I could maybe make an SVG in memory, which would help me make the crazy picklable Py5Shape objects that I want...

**Comment by villares** [comment]
More context... I was extending pymunk (2D physics  simulation engine) objects with some Py5Shape drawable content in a field... then I wanted to save/dump and reload the whole simulation with pickle and I had to take a step back and undo my simple & efficient Py5Shape based field in order to have it all "picklable".

**Comment by hx2A** [comment]
> If `begin_record()` could accept a "file-like" buffer object instead of a string, I could maybe make an SVG in memory, which would help me make the crazy picklable Py5Shape objects that I want...

This is a good idea! Doesn't `begin_record()` take a Py5Graphics object as a parameter? Can't you use `create_graphics()` to create a SVG graphics object, pass it to `begin_record()`, and create an SVG file that way? I think that would accomplish what you want.

**Reply by villares** [reply]
I thought about it,  then I thought it cumbersome because I would have to find the shape size to instatiate the PyGraphics and I forgot about it! Will try!

**Reply by hx2A** [reply]
> I thought about it, then I thought it cumbersome because I would have to find the shape size to instatiate the PyGraphics and I forgot about it! Will try!

With the SVG renderer, does it matter if the size is too big or too small? It might matter for positioning with your current `shape_mode()` but I don't think it crops if you draw outside the set size.

**Reply by villares** [reply]
And I'm getting this:
```python
>>> shp.get_height()
0.0
```

**Reply by hx2A** [reply]
I vaguely remember looking at the source code for this aspect of PShape objects and it is a bit funky. I don't think it can be changed without breaking things for others, but currently it isn't as reliable.

**Reply by villares** [reply]
I have just tested, and I'm afraid that `create_graphics()` when using the SVG renderer requires a string-or-path output argument :(

```
java.lang.RuntimeException: java.lang.RuntimeException: setOutput() or setPath() must be used with the SVG renderer
```

**Reply by hx2A** [reply]
Right. But I thought that was what you wanted? To save the shape to a file of some kind?

**Reply by villares** [reply]
No, I want to use a `io.BytesIO` file-like object to have it serialized in memory...

**Reply by hx2A** [reply]
I see. I can't think of a way to do that besides writing out the SVG file and then reading it back.

Slightly off topic, but `save()` and `save_frame()` both support writing to `io.BytesIO` because py5 uses PIL to save images, not the Processing function.
